In [145]:
import pandas as pd

df=pd.read_csv('fruit.csv', index_col = 0)

y = df['class'].values
X = df.drop(columns = 'class').values

df.head()

mass  width  height  color_score  class
0   192    8.4     7.3         0.55      1
1   180    8.0     6.8         0.59      1
2   176    7.4     7.2         0.60      1
3    86    6.2     4.7         0.80      2
4    84    6.0     4.6         0.79      2

In [146]:
# Check for NaN value
df.isnull().sum()

mass           0
width          0
height         0
color_score    0
class          0
dtype: int64

### Train-test split

In [147]:
# perform train - test split. (40% as testing and 60% as training)
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.4, random_state =1)

### KNN

In [148]:
#--------------------------------------------------
##------------ K-NN Classifier ------------------## 
#--------------------------------------------------
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier

steps = [('scaler', StandardScaler()),         
         ('knn', KNeighborsClassifier(n_neighbors = 5))]

knn_pipeline = Pipeline(steps)
knn_pipeline.fit(X_train, y_train)

Pipeline(steps=[('scaler', StandardScaler()), ('knn', KNeighborsClassifier())])

### Logistic Regression

In [149]:
# #--------------------------------------------------
# ## ----------- Logistic Regression ---------------## 
# #--------------------------------------------------
# from sklearn.preprocessing import StandardScaler 
# from sklearn.linear_model import LogisticRegression 
# from sklearn.pipeline import Pipeline

# steps = [('scaler', StandardScaler()),
#          ('logReg', LogisticRegression())]
         
# clf_pipeline = Pipeline(steps) 
# clf_pipeline.fit(X_train, y_train)

In [150]:
# check dimension of y_test
y_test

array([4, 1, 3, 4, 3, 3, 3, 1, 1, 4, 4, 2, 3, 3, 3, 1, 4, 3, 4, 3, 3, 4,
       3, 4])

In [151]:
# check dimension of ypred_testP
ypred_testP

array([[0. , 0. , 0. , 1. ],
       [1. , 0. , 0. , 0. ],
       [0.4, 0. , 0.6, 0. ],
       [0.2, 0. , 0.2, 0.6],
       [0.2, 0. , 0.8, 0. ],
       [0.8, 0. , 0.2, 0. ],
       [0.2, 0. , 0.8, 0. ],
       [0.8, 0. , 0.2, 0. ],
       [0.6, 0. , 0.4, 0. ],
       [0. , 0. , 0. , 1. ],
       [0.2, 0. , 0.2, 0.6],
       [0. , 0.8, 0.2, 0. ],
       [0.4, 0. , 0.6, 0. ],
       [0.2, 0. , 0.8, 0. ],
       [0. , 0. , 0.6, 0.4],
       [1. , 0. , 0. , 0. ],
       [0. , 0. , 0. , 1. ],
       [0.2, 0. , 0.6, 0.2],
       [0. , 0. , 0. , 1. ],
       [0.2, 0. , 0.8, 0. ],
       [0.4, 0. , 0.6, 0. ],
       [0.2, 0. , 0.2, 0.6],
       [0.2, 0. , 0.4, 0.4],
       [0.2, 0. , 0.2, 0.6]])

In [152]:
#--------------------------------------------------
## Model Evaluation ## 
#--------------------------------------------------
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_auc_score

# Training

# ypred_test = knn_pipeline.predict(X_train)
# mat_clf = confusion_matrix(y_train, ypred_test)
# report_clf = classification_report(y_train, ypred_test)

# print(mat_clf)
# print(report_clf)

# ypred_testP = knn_pipeline.predict_proba(X_train)
# auc = roc_auc_score(y_train, ypred_testP, multi_class='ovr')
# print(auc)


# Testing

ypred_test = knn_pipeline.predict(X_test)
mat_clf = confusion_matrix(y_test, ypred_test)
report_clf = classification_report(y_test, ypred_test)

print(mat_clf)
print(report_clf)

ypred_testP = knn_pipeline.predict_proba(X_test)
auc = roc_auc_score(y_test, ypred_testP, multi_class='ovr')
print(auc)

[[ 4  0  0  0]
 [ 0  1  0  0]
 [ 1  0 10  0]
 [ 0  0  0  8]]
              precision    recall  f1-score   support

           1       0.80      1.00      0.89         4
           2       1.00      1.00      1.00         1
           3       1.00      0.91      0.95        11
           4       1.00      1.00      1.00         8

    accuracy                           0.96        24
   macro avg       0.95      0.98      0.96        24
weighted avg       0.97      0.96      0.96        24

0.9874453671328671


**Accuracy for KNN (K value = 5)**

*   Training: accuracy 0.91
*   Testing: accuracy 0.96

**Accuracy for Logistic Regression (no hyperparameter tuning)**

*   Training: accuracy 0.80
*   Testing: accuracy 0.67





